[iOS 16] 'Select All' appears in the edit menu after selecting a text range

whsieh opened the pull request:
#### dd3c292ca93c8499aab292ab26e6b71fcf5665b1
<pre>
[iOS 16] &apos;Select All&apos; appears in the edit menu after selecting a text range
<a href="https://bugs.webkit.org/show_bug.cgi?id=243136">https://bugs.webkit.org/show_bug.cgi?id=243136</a>

Reviewed by Ryosuke Niwa.

In iOS 16, when presenting the edit menu (either by right clicking with a trackpad or after
selecting text via gestures), the `sender` argument that&apos;s passed to `-canPerformAction:withSender:`
is no longer the shared `UIMenuController` instance. This breaks our existing logic to suppress
&quot;Select All&quot; in the callout bar when text is already selected. Furthermore, the `sender` in this
case is just a `UIKeyCommand`, making it indistinguishable from the `sender` in the case where the
user is using `⌘+A` with a hardware keyboard to insert text.

To fix this on iOS 16, instead of consulting the value of `sender`, we set a flag when the edit menu
is about to be presented (either via right click with trackpad, or via text interaction gestures)
and check the value of this flag instead.

Covered by augmenting an existing layout test.

* LayoutTests/editing/selection/ios/show-callout-bar-after-selecting-word-expected.txt:
* LayoutTests/editing/selection/ios/show-callout-bar-after-selecting-word.html:

Augment a layout test that already checks for the presence of &quot;Copy&quot; after making a text selection,
by additionally verifying that the &quot;Select All&quot; action is not also present.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):

Consult the new flag (`_isPreparingEditMenu`) when deciding whether to allow &quot;Select All&quot;.

(-[WKContentView requestRVItemInSelectedRangeWithCompletionHandler:]):

Set `_isPreparingEditMenu` when invoking the completion handler, right before showing the callout
bar after Reveal items have been requested. This is the final step before UIKit proceeds to show the
callout bar after selecting text via a long press or loupe gesture (querying
`-canPerformAction:withSender:` in the process).

(-[WKContentView prepareSelectionForContextMenuWithLocationInView:completionHandler:]):

Set `_isPreparingEditMenu` after preparing for the (editing) context menu. Similarly, this is the
final step before UIKit proceeds to show the edit menu after right clicking to select text, querying
`-canPerformAction:withSender:` in the process.

Canonical link: <a href="https://commits.webkit.org/252764@main">https://commits.webkit.org/252764@main</a>
</pre>

rniwa commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -7695,8 +7697,13 @@ - (BOOL)supportsTextSelectionWithCharacterGranularity
+        if (!strongSelf)
+            return completionHandler(nil);
+
+        SetForScope editMenuPresentationScope { strongSelf->_isStartingToPresentEditMenu, YES };
```
isStartingToPresentEditMenu doesn't sound quite right but I don't have a better suggestion.

whsieh replied:
Yeah, it's difficult to come up with a name for this one. I do think "edit menu" is the correct term that refers to the right click edit menu and (what was previously referred to as) the callout bar, but maybe the `isStartingToPresent` part is weird.

Perhaps, something along the lines of `_shouldAvoidSelectAllForSelectedText` that instead describes how the flag is utilized?

(Also, I'll need to move the flag in order to fix the watchOS/tvOS builds)

rniwa replied:
Hm... maybe something like _isPresentingEditMenu or _isPreparingEditMenu might make sense?

whsieh replied:
Sounds good to me — I'll go with `_isPreparingEditMenu`.

rniwa reviewed (commented):
r=me

whsieh committed:
[iOS 16] 'Select All' appears in the edit menu after selecting a text range
https://bugs.webkit.org/show_bug.cgi?id=243136

Reviewed by Ryosuke Niwa.

In iOS 16, when presenting the edit menu (either by right clicking with a trackpad or after
selecting text via gestures), the `sender` argument that's passed to `-canPerformAction:withSender:`
is no longer the shared `UIMenuController` instance. This breaks our existing logic to suppress
"Select All" in the callout bar when text is already selected. Furthermore, the `sender` in this
case is just a `UIKeyCommand`, making it indistinguishable from the `sender` in the case where the
user is using `⌘+A` with a hardware keyboard to insert text.

To fix this on iOS 16, instead of consulting the value of `sender`, we set a flag when the edit menu
is about to be presented (either via right click with trackpad, or via text interaction gestures)
and check the value of this flag instead.

Covered by augmenting an existing layout test.

* LayoutTests/editing/selection/ios/show-callout-bar-after-selecting-word-expected.txt:
* LayoutTests/editing/selection/ios/show-callout-bar-after-selecting-word.html:

Augment a layout test that already checks for the presence of "Copy" after making a text selection,
by additionally verifying that the "Select All" action is not also present.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):

Consult the new flag (`_isPreparingEditMenu`) when deciding whether to allow "Select All".

(-[WKContentView requestRVItemInSelectedRangeWithCompletionHandler:]):

Set `_isPreparingEditMenu` when invoking the completion handler, right before showing the callout
bar after Reveal items have been requested. This is the final step before UIKit proceeds to show the
callout bar after selecting text via a long press or loupe gesture (querying
`-canPerformAction:withSender:` in the process).

(-[WKContentView prepareSelectionForContextMenuWithLocationInView:completionHandler:]):

Set `_isPreparingEditMenu` after preparing for the (editing) context menu. Similarly, this is the
final step before UIKit proceeds to show the edit menu after right clicking to select text, querying
`-canPerformAction:withSender:` in the process.

Canonical link: https://commits.webkit.org/252764@main
